### ISSUE TITLE
JavaScript library

### ISSUE THREAD

**Issue by xaur** [xaur/decred-issues#115]
JavaScript library was asked about multiple times.

Most recent discussion was about Edge Wallet, ElectrumX and a missing JS lib. What is needed is "address stuff and signature stuff".

Discussions:

* [2019-02-12](https://matrix.to/#/!MIGqWXfLFBwhipPKYL:decred.org/$15500054718437GyoNW:decred.org)

**Comment by xaur** [comment]
@jzbz feel free to drop any details here, I'll add them to the issue description.

**Comment by xaur** [comment]
Hey @oregonisaac I think about adding contents of your [gist](https://gist.github.com/oregonisaac/e41826e171b53826ff869cab75d8ed0b) to this issue, but wonder where did you take the first part from (Basic Coding Requirements)?

**Comment by oregonisaac** [comment]
I looked up articles on best practices and references one Typescript the site Dave referenced plus a number of JS best practice sites (did not make a reference list). 

Would be good to verify with a few folks.  I think this is what Dave was asking for -  https://www.typescriptlang.org

**Comment by jzbz** [comment]
Yeah that's the TypeScript Dave had asked for. Might also make sense to include this:
https://github.com/decred/dcrd/blob/master/docs/code_contribution_guidelines.md

All the parts from section 4 onward (not relating to Go) would still apply.

**Comment by oregonisaac** [comment]
I’ll modify the gist accordingly.

**Comment by oregonisaac** [comment]
Updated, I might make sense to fork the Go based contribution requirements for other languages and include links to the language specific standards we want followed (for example I could like to the documentation page for TypeScript and make a JS version of that page).

**Comment by jzbz** [comment]
@oregonisaac that's actually a great idea. You might want to touch bases with @s-ben as he was working on dev docs IIRC.

**Comment by s-ben** [comment]
@oregonisaac @jzbz , I think we should at least add JS guidance to the [Source Code Contributions](https://docs.decred.org/contributing/source-code-contributions/) page. Have created an [Issue](https://github.com/decred/dcrdocs/issues/862) for that on dcrdocs. 

I also think the general code contribution guidelines from dcrd are good. They're pretty specific to Go and the dcrd repo though (e.g. their GitHub process). I think it would be useful to have a page that captures our general code contribution guidelines at a higher level, but also directing people to the individual repo READMEs for specific guidelines. Have put this on my to do list.

**Comment by xaur** [comment]
Pasting [comment](https://gist.github.com/oregonisaac/e41826e171b53826ff869cab75d8ed0b#gistcomment-2840058) from @paullinator from the gist:

> I would add that the library should be platform agnostic and work under browsers and nodejs. Ideally, the library should optionally accept some crypto routines as injections to allow the wallet to provide accelerated versions of those routines. ie secp256k1, sha256, pbkdf2, and random(). All of these should accept async function injections as native code accelerated functions have to be async in most implementations.
> 
> In addition to just creating and signing transactions, I would add parsing of transactions as well.
> 
> Also, since most wallets use pre-existing bitcoin libraries for all transaction parsing and creation, I would make sure that the part of Decred that differs from bitcoin be usable as a simple standalone module. ie. some currencies such as Groestlecoin differ by only the hashing algo used for addresses and transaction creation. That hashing algo is made available as a standalone function vs having to use an entire Groestlecoin library to create a tx.

**Comment by xaur** [comment]
@JoeGruffins wrote some javascript for working with Decred transactions: https://github.com/trezor/trezor-utxo-lib/pull/9

cc @svitekpavel
